static/js/compose_actions.js

```diff
@@ -196,6 +197,68 @@ function same_recipient_as_before(msg_type, opts) {
               opts.private_message_recipient === compose_state.recipient());
 }
 
+exports.add_placeholder_text = function (opts) {
+    var placeholder_text;
+
+    if (opts.message_type === 'private') {
+        // IF PRIVATE
+        if (opts.private_message_recipient) {
+            // RECIPIENTS EXIST
+            var recipient_list = opts.private_message_recipient.split(",");
+
+            if (recipient_list.length > 1) {
+                // GROUP PM
+                var recipient_names = [];
+                var num_users = 0;
+
+                recipient_list.forEach(recipient => {
+                    var user = people.get_by_email(recipient);
+                    recipient_names[num_users] = user.full_name;
+                    num_users += 1;
+                });
+
+                placeholder_text = "Message ";
+
+                recipient_names.forEach(recipient_name => {
+                    num_users = num_users - 1;
+                    if (num_users > 0) {
+                        placeholder_text += recipient_name + ", ";
+                    } else {
+                        placeholder_text += recipient_name;
+                    }
+                });
+            } else {
+                // ONE-ON-ONE PM
+                var user = people.get_by_email(recipient_list[0]);
+                var recipient_name = user.full_name;
+                var status = user_status.get_status_text(user.user_id);
+
+                if (status) {
+                    placeholder_text = "Message " + recipient_name + " (" + status + ")";
+                } else {
+                    placeholder_text = "Message " + recipient_name;
+                }
+            }
+        } else {
+            // NO RECIPIENTS
+            placeholder_text = "Compose your message here";
+        }
+    } else if (opts.message_type === 'stream') {
+        if (opts.topic) {
+            // STREAM AND TOPIC NAMES EXIST
+            placeholder_text = "Message #" + opts.stream + " > " + opts.topic;
+        } else if (opts.stream) {
+            // ONLY STREAM NAME EXISTS
+            placeholder_text = "Message #" + opts.stream;
+        } else {
+            // NEITHER STREAM NAME OR TOPIC NAME EXIST
+            placeholder_text = "Compose your message here";
+        }
+    }
+
+    $('#compose-textarea').attr('placeholder', placeholder_text);
+};
```

Comment:
Woah, this is a ton of code.  I think a good structure would be to have a `compute_placeholder_text` function in `compose_ui.js` that we can unit test for the various corner cases.

It also seems likely to me that we can do some refactoring to this function to make it nicer, but we should address the above first.

Reply:
Moved that code to `compose_ui.js` and added the tests for it. I agree that `compute_placeholder_text` is definitely pretty "ugly" at the moment. I would love to get some feedback on how I can refactor this.